### PR TITLE
API for shared service templates

### PR DESCRIPTION
**Pull request by tanya-borisova**
# PR for issue https://github.com/microsoft/AzureTRE/issues/1238

## What is being addressed

Added API for adding, listing and retrieving shared service templates. 

## How is this addressed

- Added new routes for `GET /shared-service-templates`,  `GET /shared-service-templates/{shared_service_template_name}`, and `POST /shared-service-templates` 
- Added unit tests for these routes
- Reused ResourceTemplateRepository for storing templates for shared services 
- In several places, refactored large Python files by splitting them into smaller files, e.g. extracted api_app/api/routes/user_resource_templates.py from api_app/api/routes/workspace_service_templates.py , and the same for their tests. I will mark these instances in the code appropriately to hopefully make the PR review easier.

### CODE REVIEW

File: api_app/api/routes/templates.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+from fastapi import HTTPException
+from starlette import status
+
+from db.errors import DuplicateEntity, EntityDoesNotExist
+from db.repositories.resource_templates import ResourceTemplateRepository
+from models.domain.resource import ResourceType
+from resources import strings
+
+
+def get_current_template_by_name(template_name: str, template_repo: ResourceTemplateRepository, resource_type: ResourceType, parent_service_template_name: str = "", is_update: bool = False) -> dict:
```
Comment (tanya-borisova) [inline api_app/api/routes/templates.py:12]: This was moved from https://github.com/microsoft/AzureTRE/pull/1457/files#diff-b82ed3b790102de55f2ccb34b09e72cb9dcea8b64418d34f4f8e85463d89e757

File: api_app/api/routes/user_resource_templates.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+from models.schemas.user_resource_template import UserResourceTemplateInResponse, UserResourceTemplateInCreate
+from models.schemas.resource_template import ResourceTemplateInformationInList
+from resources import strings
+from services.authentication import get_current_admin_user, get_current_tre_user_or_tre_admin
+
+
+user_resource_templates_core_router = APIRouter(dependencies=[Depends(get_current_tre_user_or_tre_admin)])
+
+
+@user_resource_templates_core_router.get("/workspace-service-templates/{service_template_name}/user-resource-templates", response_model=ResourceTemplateInformationInList, name=strings.API_GET_USER_RESOURCE_TEMPLATES, dependencies=[Depends(get_current_tre_user_or_tre_admin)])
```
Comment (tanya-borisova) [inline api_app/api/routes/user_resource_templates.py:20]: These were moved from https://github.com/microsoft/AzureTRE/pull/1457/files#diff-720854f32cb2a9f5d1d95fdb7fbf5d40c94e5671e76ba0986101244cad6c9b78

File: api_app/tests_ma/test_api/test_routes/test_user_resource_templates.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+from services.authentication import get_current_admin_user, get_current_tre_user_or_tre_admin
+from db.errors import DuplicateEntity, EntityDoesNotExist, EntityVersionExist, UnableToAccessDatabase
+from models.domain.resource import ResourceType
+from models.domain.user_resource_template import UserResourceTemplate
+from models.schemas.resource_template import ResourceTemplateInformation
+from resources import strings
+
+pytestmark = pytest.mark.asyncio
+
+
```
Comment (tanya-borisova) [inline api_app/tests_ma/test_api/test_routes/test_user_resource_templates.py:16]: These were moved from https://github.com/microsoft/AzureTRE/pull/1457/files#diff-490bc9e46273028df440ce4b5aaa0ec2bc4786117f4e7eaf2e914746d7cec4aa

File: api_app/tests_ma/test_db/test_repositories/test_user_resource_templates_repository.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
+import pytest
+from mock import patch
+
+from db.errors import DuplicateEntity, EntityDoesNotExist
+from models.domain.resource import ResourceType
+from models.domain.user_resource_template import UserResourceTemplate
+from .test_resource_templates_repository import resource_template_repo
+
+
```
Comment (tanya-borisova) [inline api_app/tests_ma/test_db/test_repositories/test_user_resource_templates_repository.py:9]: These tests were moved from https://github.com/microsoft/AzureTRE/pull/1457/files#diff-c13922086d5fcd76731b03e1d0a6b2681036ccdff2fa34a9bf40fb9d9d4dfa1e

**Review by tanya-borisova — COMMENTED** [review]
Clarifying some changes made in the PR

**Review by ross-p-smith — APPROVED** [review]

File: api_app/schemas/shared_service.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  "description": "These parameters are required for all shared services",
+  "required": [
+      "display_name",
+      "description"
+  ],
+  "properties": {
+      "display_name": {
+          "type": "string",
+          "title": "Name for the shared service",
+          "description": "The name of the shared service to be displayed to users"
```
Comment (damoodamoo) [inline api_app/schemas/shared_service.json:15]: Don't worry about for this PR - but we could probably add `updateable: true` for display name and description.

**Review by damoodamoo — APPROVED** [review]
LGTM!

File: api_app/schemas/shared_service.json (hunk 5 above)
Comment (tanya-borisova) [reply api_app/schemas/shared_service.json:15]: Added a note for my next PR!

**Pull request merged**